### ISSUE TITLE
Prevent Duplicate Records

### ISSUE THREAD

**Issue by philipbjorge** [huddlej/pnwmoths#243]
Currently duplicate SpeciesRecords can be added to the database.
You'll want to prevent this by adding a unique_together constraint to the SpeciesRecord model.

```
unique_together = (("driver", "restaurant"),)
```

You should probably add every field except id.

**Comment by MatteBailey** [comment]
By id do you mean the "species" field? Also, why do we want to leave this out?

**Comment by MatteBailey** [comment]
I got this working using unique_together with the following fields:

```
unique_together = ("record_type", "species", "latitude", "longitude", "year", "month", "day")
```

I originally had more, but it failed to prevent duplicates when I tested it. I think this is because certain fields might be left empty or be slightly different from the original in some undetectable way. I think these fields should be adequate though, as I can't think of a circumstance where someone would intentionally upload a species record that has the same voucher type, species type, location data, and date as one that already exists.

Do you have any issues with this before I push it?

**Comment by philipbjorge** [comment]
So here are the properties of a SpeciesRecord (from models.py):

```
    record_type = models.CharField(max_length=20, choices=RECORD_TYPE_CHOICES, verbose_name='Voucher Type')
    species = models.ForeignKey(Species)
    type_status = models.CharField(null=True, blank=True, max_length=255)
    latitude = models.FloatField(null=True, blank=True)
    longitude = models.FloatField(null=True, blank=True)
    locality = models.CharField(null=True, blank=True, max_length=255)
    county = models.ForeignKey(County, null=True, blank=True)
    state = models.ForeignKey(State, null=True, blank=True)
    elevation = models.IntegerField(help_text="measured in feet", null=True, blank=True)
    year = models.IntegerField(null=True, blank=True)
    month = models.IntegerField(null=True, blank=True, validators=[MinValueValidator(1), MaxValueValidator(12)])
    day = models.IntegerField(null=True, blank=True, validators=[MinValueValidator(1), MaxValueValidator(31)])
    collector = models.ForeignKey(Collector, null=True, blank=True)
    collection = models.ForeignKey(Collection, null=True, blank=True)
    males = models.IntegerField(null=True, blank=True, help_text=gender_docs)
    females = models.IntegerField(null=True, blank=True, help_text=gender_docs)
    notes = models.TextField(null=True, blank=True)

    """
    The following should not be included in the unique_together. If they were, duplicate records would only
    be prevented if they were uploaded at the exact same time (basically impossible) and/or with the same CSV
    filename.
    """
    csv_file = models.CharField(null=True, blank=True, max_length=255)
    date_added = models.DateTimeField(editable=False)
    date_modified = models.DateTimeField(editable=False)
```

I'm pretty sure when you talk to Merrill you'll find there are enough special cases that you'll need to have every property being checked.

That being said, I could see there being an issue with using unique_together on so many properties - try googling for something like unique_together property limit. Were you included date_added, date_modified, or csv_file? Because that would cause duplicates to never be found.

**Comment by philipbjorge** [comment]
And by id, there is an automatic incrementing id added to each model by Django for uniquely identifying each record. You can see it in the admin URLs (e.g. admin/speciesrecord/12345)

Including this in unique together would cause the same problem as including date_added. It would change on each upload never catching duplicates.

**Comment by MatteBailey** [comment]
I think I might have had date_added in there before, good call. I added some more fields, and have it working with this:

```
unique_together = ("record_type", "species", "latitude", "longitude", "locality", "county", "state", "elevation", "year", "month", "day", "collector", "collection", "males", "females", "notes")
```

I also had type_status in there previously, but it didn't prevent duplicates when I tested it for some reason. This field is blank in all the records I've checked so far, so I wouldn't think leaving it out would be too big of a concern.
